contrib/lib/src/databases.rs

```diff
@@ -0,0 +1,148 @@
+pub extern crate r2d2;
+
+use std::collections::BTreeMap;
+use std::io;
+use std::marker::{Send, Sized};
+use rocket::config::{self, ConfigError, Value};
+
+use self::r2d2::ManageConnection;
+
+#[cfg(feature = "diesel")]
+pub extern crate diesel;
+
+#[cfg(feature = "pg_pool")]
+pub extern crate postgres;
+#[cfg(feature = "pg_pool")]
+pub extern crate r2d2_postgres;
+
+#[cfg(feature = "redis_pool")]
+pub extern crate redis;
+#[cfg(feature = "redis_pool")]
+pub extern crate r2d2_redis;
+
+/// A struct containing database configuration options.
+///
+/// This retrieves the key-value pairs specified in the `Rocket.toml` under
+/// the databases key for use in further configuring the desired connection.
+pub struct DatabaseConfig<'a> {
+    pub url: &'a str,
+    pub pool_size: u32,
+    pub extras: Option<&'a BTreeMap<String, Value>>,
+}
+
+/// A custom error type to wrap multiple, disparate error types
+///
+/// This allows for returning the given error from the connection manager
+/// creation methods and the r2d2 connection pool methods.
+pub enum DbError<T> {
+    Custom(T),
+    PoolError(r2d2::Error),
+}
+
+/// This method retrieves the database configuration from the `Rocket.toml` file
+/// and returns a DatabaseConfig struct.
+pub fn database_config<'a>(name: &str, from: &'a config::Config) -> Option<DatabaseConfig<'a>> {
```

Comment:
I still think we can improve this method considerably.

Reply:
I agree. I'm just not prioritizing it at this point in time. I figured I'd take another pass over it as we got closer to merging this PR.